doc/dev/static_type_checking_cheat_sheet.md

```diff
@@ -0,0 +1,410 @@
+from typing import Any
+
+# args seen as tuple[str] by type checker
+# kwargs seen as dict[str, Any] by type checker
+def begin_operation(*args: str, **kwargs: Any) -> None:
+    ...
+```
+
+### Structural subtyping / Protocols
+
+- Do support the "duck typing" of Python with `typing.Protocol`.
+- Mark your `Protocol`s as `@runtime_checkable` so users can use them in `isinstance` and `issubclass` checks.
+
+```python
+from typing_extensions import Protocol, runtime_checkable
+
+@runtime_checkable
+class TokenCredential(Protocol):
+    """Protocol for classes able to provide OAuth tokens."""
+
+    def get_token(
+        self, *scopes: str, claims: Optional[str] = None, tenant_id: Optional[str] = None, **kwargs: Any
+    ) -> AccessToken:
+        ...
+```
+
+### Generics
+
+- Use `typing.TypeVar` to let functions/classes work with arguments of any type and retain relationships between arguments and their return values.
+
+```python
+from typing import TypeVar, Sequence, Any
+
+# No:
+def choose(items: Sequence[Any]) -> Any:
+    ...
+
+# Yes:
+T = TypeVar("T")
+
+def choose(items: Sequence[T]) -> T:
+    ...
+```
+
+- Try to constrain your `TypeVar`'s by specifying types or using the `bound` keyword argument.
+
+```python
+from typing import TypeVar
+
+T = TypeVar("T", int, str)  # limited to int or str
+S = TypeVar("S", bound=str)  # limited to str or any subtype of str
+```
+
+- Do subclass `typing.Generic` or a generic collection type if your class needs to work with multiple types.
+- Do give your `typing.TypeVar`'s descriptive names if they will be publicly exposed in the code.
+
+```python
+from typing import TypeVar, Generic
+
+# No:
+T = TypeVar("T")
+
+class LROPoller(Generic[T]):
+    ...
+
+# Yes:
+PollingReturnType = TypeVar("PollingReturnType")
+
+class LROPoller(Generic[PollingReturnType]):
+    ...
+```
+
+- Follow naming conventions for covariant (`*_co`) and contravariant (`*_contra`) `TypeVar` parameters when setting variance for generic classes.
+
+```python
+from typing import TypeVar
+
+T_co = TypeVar("T_co", covariant=True)
+T_contra = TypeVar("T_contra", contravariant=True)
+```
+
+- Use the pre-defined TypeVar, `typing.AnyStr`, when your parameter or return type expects both `str` or `bytes`.
+- Do use `typing.ParamSpec` (a specialized TypeVar) to forward your function type hints from one callable to another callable (e.g. decorators). This preserves type hinting in the original function.
+
+```python
+from typing import TypeVar, Callable, ParamSpec
+
+
+T = TypeVar("T")
+P = ParamSpec("P")
+
+def validate_table(func: Callable[P, T]) -> Callable[P, T]:
+    def inner(*args: P.args, **kwargs: P.kwargs) -> T:
+        # validate table
+        return func(*args, **kwargs)
+    return inner
+
+
+@validate_table
+def create_table(table_name: str, **kwargs: Any) -> None:
+    ...
+```
+
+
+### Type Aliases
+
+- Do use type aliases to help make lengthy, complicated type hints more readable or help convey meaning to the reader.
```

Comment:
I think it's worth making a distinction here between the type hint and the docstring parameter type.
Currently - if we're using type aliases, they should be fully expanded in the docstring. At least, last I checked, using the type alias in the docstring didn't render in a useful way in sphinx.

Reply:
Question: If using a type alias in a public API - does the alias itself need to be part of the public API?
I imagine not for use of a static type checker, but what about rendered docs?

Reply:
Good point - making that distinction. Yeah type alias in the docstring is no good with sphinx.

> If using a type alias in a public API - does the alias itself need to be part of the public API? I imagine not for use of a static type checker, but what about rendered docs?

I don't believe it needs to be public API. edit: tested it and in the rendered docs it expands the type alias to the actual type so we're good here.